### PR TITLE
[SPARK-38918][SQL][3.1] Nested column pruning should filter out attributes that do not belong to the current relation

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

Backport https://github.com/apache/spark/pull/36216 to branch-3.1.

### Why are the changes needed?

To fix a bug in `SchemaPruning`.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by viirya** [comment]
Seems related test failure?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 1]
```diff
@@ -597,8 +597,7 @@ abstract class SchemaPruningSuite
             |  and e.employer.name = c.employer.company.name)
             |""".stripMargin)
         checkScan(query,
-          "struct<name:struct<first:string,middle:string,last:string>," +
-            "employer:struct<id:int,company:struct<name:string,address:string>>>",
+          "struct<name:struct<first:string>,employer:struct<company:struct<name:string>>>",
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:600]: Hmm, pruned schema is different in 3.1?

Comment (allisonwang-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:600]: Yes looks like starting from 3.2 it failed to prune some nested fields (for example in this case `name.middle` and `name.last` are not used)

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala:600]: Hmm, okay, but I think it is unrelated to this change.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks. Merging to branch-3.1.

**Pull request closed without merging**